### PR TITLE
[gstreamer] create a new port

### LINKED ISSUE #1474: Add gstreamer
It would be great to have a port of the gstreamer library ( https://gstreamer.freedesktop.org/ ).

Debian package : https://packages.debian.org/source/sid/gstreamer1.0 .
Homebrew formula : https://github.com/Homebrew/homebrew-core/blob/master/Formula/gstreamer.rb .

### LINKED ISSUE #6092: help add repo  gstreamer and it's various plugins
gstreamer is strong and powerfull, many open source depend on it. but it is complex  to port。

### LINKED ISSUE #7958: gstreamer not found
i couldn't find gstreamer in vcpkg list.

### PR DESCRIPTION
**Pull request by luncliff**
### What does your PR fix?  

> I'm learning meson so this work can take a long time. Every opinion will be helpful!

Resolves #1474 
Resolves #6092
Resolves #7958

Thanks to #13100 I was able to build `gstreamer` with `meson`.

I'm not a `meson` user, but a bit interested. Suggestions will be welcomed.
Currently considering 3 features with gstreamer plugins. But, only 2 will be available in this PR.

* `gstreamer[plugins-good]`: default feature
* `gstreamer[plugins-ugly]`: default feature
* `gstreamer[plugins-bad]`: Not supported. Will be `FATAL_ERROR`

#### Related Issues?

* #6683 (Dependency handling with `meson`)
* #6620 (`opencv` + `gstreamer`): Not considered in this PR

### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  

Only a few triplets will be supported.

* `x64-windows`
* `x86-windows`
* `x64-osx`

### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  

Yes.

Especially, I'm considering the separation of `vcpkg_configure_meson` related changes for this work.  
I will do the work when the port build becomes more stable. (bdc2595, 8a67fbd)

### CODE REVIEW

**Comment by luncliff** [comment]
#13100 makes too many changes for review. Windows triplets are still WIP so I will leave it unmerged until then.
Will be rebased after the merge of  #13100.

**Comment by luncliff** [comment]
In Windows, The gstreamer projects' meson configuration finds `z.lib` instead of `zlib.lib`. 
`plugins-base-use-zlib.patch` and `plugins-good-use-zlib.patch` didn't solve the problem. :( 

https://github.com/mesonbuild/meson/pull/6421#issuecomment-575769595 looks helpful so I'm gonna try it soon.

**Comment by luncliff** [comment]
By the way, `yasm-tool:x86-windows` need to be installed for `gstreamer:x64-windows`. I added `yasm-tool-helper` to make error message more understandable.

Can we specify `triplets` in [`dependencies`](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/manifest-files.md#dependencies)? Or something similar?  
At a glance, it seems like host dependencies are for different purpose.

File: ports/gstreamer/plugins-base-use-zlib.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/gst-libs/gst/tag/meson.build b/gst-libs/gst/tag/meson.build
+index 5ec3739..d2378c6 100644
+--- a/gst-libs/gst/tag/meson.build
++++ b/gst-libs/gst/tag/meson.build
+@@ -72,7 +72,7 @@ core_conf.set('HAVE_ISO_CODES', have_iso_codes)
+ # could drop optional zlib dep and use g_zlib_decompressor_new()
+ zlib_dep = dependency('zlib', required : false)
```
Comment (Neumann-A) [inline ports/gstreamer/plugins-base-use-zlib.patch:7]:
```suggestion
 zlib_dep = dependency('zlib', required : true)
```

File: ports/gstreamer/plugins-good-use-zlib.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/meson.build b/meson.build
+index 9de3b01..d37b8ad 100644
+--- a/meson.build
++++ b/meson.build
+@@ -328,7 +328,7 @@ endif
+ 
+ zlib_dep = dependency('zlib', required : false)
```
Comment (Neumann-A) [inline ports/gstreamer/plugins-good-use-zlib.patch:7]:
```suggestion
 zlib_dep = dependency('zlib', required : true)
```

File: ports/gstreamer/plugins-base-use-zlib.patch [hunk 3]
```diff
@@ -0,0 +1,13 @@
+diff --git a/gst-libs/gst/tag/meson.build b/gst-libs/gst/tag/meson.build
+index 5ec3739..d2378c6 100644
+--- a/gst-libs/gst/tag/meson.build
++++ b/gst-libs/gst/tag/meson.build
+@@ -72,7 +72,7 @@ core_conf.set('HAVE_ISO_CODES', have_iso_codes)
+ # could drop optional zlib dep and use g_zlib_decompressor_new()
+ zlib_dep = dependency('zlib', required : false)
+ if not zlib_dep.found()
+-  zlib_dep = cc.find_library('z', required : false)
++  zlib_dep = cc.find_library('zlib', required : true)
```
Comment (Neumann-A) [inline ports/gstreamer/plugins-base-use-zlib.patch:10]: this change is not necessary

File: ports/gstreamer/plugins-good-use-zlib.patch [hunk 4]
```diff
@@ -0,0 +1,13 @@
+diff --git a/meson.build b/meson.build
+index 9de3b01..d37b8ad 100644
+--- a/meson.build
++++ b/meson.build
+@@ -328,7 +328,7 @@ endif
+ 
+ zlib_dep = dependency('zlib', required : false)
+ if not zlib_dep.found()
+-  zlib_dep = cc.find_library('z', required : false)
++  zlib_dep = cc.find_library('zlib', required : true)
```
Comment (Neumann-A) [inline ports/gstreamer/plugins-good-use-zlib.patch:10]: same as above

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 5]
```diff
@@ -328,6 +328,10 @@ function(vcpkg_configure_meson)
     debug_message("Including cmake vars from: ${_VCPKG_CMAKE_VARS_FILE}")
     include("${_VCPKG_CMAKE_VARS_FILE}")
 
+    vcpkg_find_acquire_program(GIT)
+    get_filename_component(GIT_DIR "${GIT}" DIRECTORY)
+    vcpkg_add_to_path("${GIT_DIR}")
+
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:334]: No. Don't give meson access to git by default please. If meson requires git for your port add it in the portfile.

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 6]
```diff
@@ -129,9 +129,9 @@ function(vcpkg_internal_meson_generate_flags_properties_string _out_var _config)
     vcpkg_internal_meson_convert_list_to_python_array(MESON_CXXFLAGS_${_config} ${MESON_CXXFLAGS_${_config}})
     string(APPEND ${_out_var} "cpp_args = ${MESON_CXXFLAGS_${_config}}\n")
     if(VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
-        set(LINKER_FLAGS_${_config} "${VCPKG_DETECTED_CMAKE_SHARED_LINKER_FLAGS_${_config}}")
+        list(APPEND LINKER_FLAGS_${_config} "${VCPKG_DETECTED_CMAKE_SHARED_LINKER_FLAGS_${_config}}")
     else()
-        set(LINKER_FLAGS_${_config} "${VCPKG_DETECTED_CMAKE_STATIC_LINKER_FLAGS_${_config}}")
+        list(APPEND LINKER_FLAGS_${_config} "${VCPKG_DETECTED_CMAKE_STATIC_LINKER_FLAGS_${_config}}")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:134]: Why?

File: ports/gstreamer/plugins-base-use-zlib.patch (hunk 3 above)
Comment (luncliff) [reply ports/gstreamer/plugins-base-use-zlib.patch]: Applied in 022f10e 👍

File: ports/gstreamer/plugins-good-use-zlib.patch (hunk 4 above)
Comment (luncliff) [reply ports/gstreamer/plugins-good-use-zlib.patch]: Applied in 022f10e 👍

File: ports/gstreamer/plugins-good-use-zlib.patch (hunk 2 above)
Comment (luncliff) [reply ports/gstreamer/plugins-good-use-zlib.patch]: Applied in 022f10e. 👍

File: ports/gstreamer/plugins-base-use-zlib.patch (hunk 1 above)
Comment (luncliff) [reply ports/gstreamer/plugins-base-use-zlib.patch]: Applied in 022f10e. 👍

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 5 above)
Comment (luncliff) [reply scripts/cmake/vcpkg_configure_meson.cmake]: reverted related changes in fd1d21a :D

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 6 above)
Comment (luncliff) [reply scripts/cmake/vcpkg_configure_meson.cmake]: reverted related changes in fd1d21a :D

**Comment by luncliff** [comment]
Sorry for late update. The port build should be successful in `x64-osx`.
Currently, I can't find where `z.lib` comes from(in Windows build).

Will be glad someone can solve this issue.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
Hi @luncliff 

Have you tested the features in this PR? I noticed that you have commented something about these features. But there is no any description about `nls`.

**Comment by luncliff** [comment]
Thanks for ping, @NancyLi1013. I added it in very early work of this PR, but now I think it won't be needed. I will remove it.

**Comment by luncliff** [comment]
Currently this port shoud success for `x64-windows`, `x86-windows`, and `x64-osx`.

**Comment by ras0219-msft** [comment]
This LGTM, thanks for the PR!

**Pull request merged**